cli/cliui/agent.go

```diff
@@ -28,6 +34,28 @@ type AgentOptions struct {
 	DocsURL       string
 }
 
+// agentWaiter encapsulates the state machine for waiting on a workspace agent.
+type agentWaiter struct {
+	opts       AgentOptions
+	sw         *stageWriter
+	logSources map[uuid.UUID]codersdk.WorkspaceAgentLogSource
+	fetchAgent func(context.Context) (codersdk.WorkspaceAgent, error)
+}
+
+// pollWhile polls the agent while the condition is true. It fetches the agent
+// on each iteration and returns the updated agent when the condition is false,
+// the context is canceled, or an error occurs.
+func (aw *agentWaiter) pollWhile(ctx context.Context, agent codersdk.WorkspaceAgent, cond func(agent codersdk.WorkspaceAgent) bool) (codersdk.WorkspaceAgent, error) {
+	var err error
+	for cond(agent) {
+		agent, err = aw.fetchAgent(ctx)
```

Comment:
nit: you don't want to handle ctx in this function (`pollWhile`)?

Reply:
`fetchAgent` already returns `ctx.Err()` when canceled, so I think we're covered. Were you thinking of something specific?

Reply:
Ahh I see, like an early cancel detection. Ya I can add that not a bad idea 👍 (improvement from original code).